cli/src/main/java/com/devonfw/tools/ide/merge/DirectoryMerger.java

```diff
@@ -0,0 +1,137 @@
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Set;
+
+import org.jline.utils.Log;
+
+import com.devonfw.tools.ide.context.IdeContext;
+import com.devonfw.tools.ide.environment.EnvironmentVariables;
+import com.devonfw.tools.ide.util.FilenameUtil;
+
+/**
+ * Implementation of {@link WorkspaceMerger} that does the whole thing:
+ * <ul>
+ * <li>It will recursively traverse directories.</li>
+ * <li>For each setup or update file if will delegate to the according {@link FileMerger} based on the file
+ * extension.</li>
+ * </ul>
+ */
+public class DirectoryMerger extends AbstractWorkspaceMerger {
+
+  private final Map<String, FileMerger> extension2mergerMap;
+
+  private final FallbackMerger fallbackMerger;
+
+  /**
+   * The constructor.
+   *
+   * @param context the {@link #context}.
+   */
+  public DirectoryMerger(IdeContext context) {
+
+    super(context);
+    this.extension2mergerMap = new HashMap<>();
+    PropertiesMerger propertiesMerger = new PropertiesMerger(context);
+    this.extension2mergerMap.put("properties", propertiesMerger);
+    this.extension2mergerMap.put("prefs", propertiesMerger); // Eclipse specific
+    XmlMerger xmlMerger = new XmlMerger(context);
+    this.extension2mergerMap.put("xml", xmlMerger);
+    this.extension2mergerMap.put("xmi", xmlMerger);
+    this.extension2mergerMap.put("launch", xmlMerger); // Eclipse specific
+    JsonMerger jsonMerger = new JsonMerger(context);
+    this.extension2mergerMap.put("json", jsonMerger);
+    this.fallbackMerger = new FallbackMerger(context);
+  }
+
+  @Override
+  public void merge(Path setup, Path update, EnvironmentVariables variables, Path workspace) {
+
+    Set<String> children = null;
+    children = addChildren(setup, children);
+    children = addChildren(update, children);
+    if (children == null) {
+      // file merge
+      FileMerger merger = getMerger(workspace);
+      merger.merge(setup, update, variables, workspace);
+    } else {
+      // directory scan
+      for (String filename : children) {
+        merge(setup.resolve(filename), update.resolve(filename), variables, workspace.resolve(filename));
+      }
+    }
+  }
+
+  private FileMerger getMerger(Path file) {
+
+    String filename = file.getFileName().toString();
+    String extension = FilenameUtil.getExtension(filename);
+    if (extension == null) {
+      this.context.debug("No extension for " + file);
+    } else {
+      this.context.trace("Extension is " + extension);
+      FileMerger merger = this.extension2mergerMap.get(extension);
+      if (merger != null) {
+        return merger;
+      }
+    }
+    return this.fallbackMerger;
+  }
+
+  @Override
+  public void inverseMerge(Path workspace, EnvironmentVariables variables, boolean addNewProperties, Path update) {
+
+    if (Files.isDirectory(update)) {
+      if (!Files.isDirectory(workspace)) {
+        Log.warn("Workspace is missing directory: " + workspace);
```

Comment:
Why are we using jline.Log instead of this.context.warning here?

Reply:
accident and mistake on my end: the old code was containing `Log.warn` and I removed out `Log` class and the static import and thought that I had replaced all these with a proper `context.warn` but missed some.
By saving the Eclipse Save Actions automatically added the import to `org.jline`.

Thank you so much for finding and fixing this "bug".